File: CS/testSample/TableBarItem/CustomBarItem/TableBarItem.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Styles;
using DevExpress.XtraBars.Controls;
using DevExpress.XtraEditors;
using DevExpress.Utils;
using DevExpress.XtraBars.ViewInfo;
using testSample.CustomBarItem;

namespace testSample
{
    public class TableBarItem : BarButtonItem
    {
        public TableBarItem()
        { 
            MenuChanged += new InitMenuChangedEventHander(TableBarItem_MenuChanged);
            InitContainer();  
        }
        public virtual void TableBarItem_MenuChanged()
        {           
            InitContainer();            
        }
        public TableBarItem(BarManager manager, string caption)
            : base(manager, caption)
        {

        }
        public TableBarItem(BarManager manager, string caption, int imageIndex)
            : base(manager, caption, imageIndex)
        {

        }
        public TableBarItem(BarManager manager, string caption, int imageIndex, BarShortcut shortcut)
            : base(manager, caption, imageIndex, shortcut)
        {

        }
        public delegate void InitMenuChangedEventHander();
        public event InitMenuChangedEventHander MenuChanged;
        public XtraUserControl InnerPopupControl
        {
            get { return _InnerPopupControl; }
            set {
                var OldValue = _InnerPopupControl;              
                if (OldValue != value)
                {
                    if (!DesignMode)
                        value.Visible = false;
                    _InnerPopupControl = value;                    
                    OnMenuChanged();
                }
            }
        }
        private XtraUserControl _InnerPopupControl;
        public delegate void OnTableClickEventHandler(object sender, TableEventArgs e);
        public event OnTableClickEventHandler TableClick;
        public TableSizePopupControlContainer container;
        public override BarButtonStyle ButtonStyle { get { return BarButtonStyle.DropDown; } set { } }
        public override PopupControl DropDownControl { get { return container; } set { } }
        public virtual void InitContainer()
        {
            if (_InnerPopupControl != null)
            {
                container = new TableSizePopupControlContainer(_InnerPopupControl);
            }
            else
            {
                container = new TableSizePopupControlContainer();
            }
            container.Click -= new EventHandler(container_Click);
            container.Click += new EventHandler(container_Click);  
        }
        void container_Click(object sender, EventArgs e)
        {
            TableSizePopupControlContainer cntr = sender as TableSizePopupControlContainer;
            if (cntr != null)
                OnTableClick(sender, new TableEventArgs(cntr.Panel.Controls[0].Text));

        }
        public virtual void OnMenuChanged()
        {
            InitMenuChangedEventHander handler = MenuChanged;
            if (handler != null)
                handler();
        }
        protected virtual void OnTableClick(object sender, TableEventArgs e)
        {
            OnTableClickEventHandler handler = TableClick;
            if (handler != null)
                handler(sender, e);
        }
     
    }
}

File: VB/testSample/TableBarItem/CustomBarItem/TableBarItem.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Styles
Imports DevExpress.XtraBars.Controls
Imports DevExpress.XtraEditors
Imports DevExpress.Utils
Imports DevExpress.XtraBars.ViewInfo
Imports testSample.CustomBarItem

Namespace testSample
    Public Class TableBarItem
        Inherits BarButtonItem

        Public Sub New()
            AddHandler Me.MenuChanged, AddressOf TableBarItem_MenuChanged
            InitContainer()
        End Sub
        Public Overridable Sub TableBarItem_MenuChanged()
            InitContainer()
        End Sub
        Public Sub New(ByVal manager As BarManager, ByVal caption As String)
            MyBase.New(manager, caption)

        End Sub
        Public Sub New(ByVal manager As BarManager, ByVal caption As String, ByVal imageIndex As Integer)
            MyBase.New(manager, caption, imageIndex)

        End Sub
        Public Sub New(ByVal manager As BarManager, ByVal caption As String, ByVal imageIndex As Integer, ByVal shortcut As BarShortcut)
            MyBase.New(manager, caption, imageIndex, shortcut)

        End Sub
        Public Delegate Sub InitMenuChangedEventHander()
        Public Event MenuChanged As InitMenuChangedEventHander
        Public Property InnerPopupControl() As XtraUserControl
            Get
                Return _InnerPopupControl
            End Get
            Set(ByVal value As XtraUserControl)
                Dim OldValue = _InnerPopupControl
                If OldValue IsNot value Then
                    If Not DesignMode Then
                        value.Visible = False
                    End If
                    _InnerPopupControl = value
                    OnMenuChanged()
                End If
            End Set
        End Property
        Private _InnerPopupControl As XtraUserControl
        Public Delegate Sub OnTableClickEventHandler(ByVal sender As Object, ByVal e As TableEventArgs)
        Public Event TableClick As OnTableClickEventHandler
        Public container As TableSizePopupControlContainer
        Public Overrides Property ButtonStyle() As BarButtonStyle
            Get
                Return BarButtonStyle.DropDown
            End Get
            Set(ByVal value As BarButtonStyle)
            End Set
        End Property
        Public Overrides Property DropDownControl() As PopupControl
            Get
                Return container
            End Get
            Set(ByVal value As PopupControl)
            End Set
        End Property
        Public Overridable Sub InitContainer()
            If _InnerPopupControl IsNot Nothing Then
                container = New TableSizePopupControlContainer(_InnerPopupControl)
            Else
                container = New TableSizePopupControlContainer()
            End If
            RemoveHandler container.Click, AddressOf container_Click
            AddHandler container.Click, AddressOf container_Click
        End Sub
        Private Sub container_Click(ByVal sender As Object, ByVal e As EventArgs)
            Dim cntr As TableSizePopupControlContainer = TryCast(sender, TableSizePopupControlContainer)
            If cntr IsNot Nothing Then
                OnTableClick(sender, New TableEventArgs(cntr.Panel.Controls(0).Text))
            End If

        End Sub
        Public Overridable Sub OnMenuChanged()
            Dim handler As InitMenuChangedEventHander = MenuChangedEvent
            If handler IsNot Nothing Then
                handler()
            End If
        End Sub
        Protected Overridable Sub OnTableClick(ByVal sender As Object, ByVal e As TableEventArgs)
            Dim handler As OnTableClickEventHandler = TableClickEvent
            If handler IsNot Nothing Then
                handler(sender, e)
            End If
        End Sub

    End Class
End Namespace
